src/onemod/pipeline.py

```diff
@@ -175,6 +209,96 @@ def stage_from_json(
             stage_class = getattr(module, stage_type)
         return stage_class.from_json(filepath, name, from_pipeline)
 
+    def build_dag(self) -> dict[str, list[str]]:
+        """Build directed acyclic graph (DAG) from the stages and their dependencies."""
+        # TODO: Placeholder until DAG class is implemented, assuming we need one
+        return self._dependencies
+
+    def validate_dag(self):
+        """Validate that the DAG structure is correct."""
+        for stage, dependencies in self._dependencies.items():
+            # Check for undefined dependencies
+            for dep in dependencies:
+                if dep not in self._stages:
+                    raise ValueError(
+                        f"Stage '{dep}' is not defined, but '{stage}' depends on it."
+                    )
+
+            # Check for self-dependencies
+            if stage in dependencies:
+                raise ValueError(f"Stage '{stage}' cannot depend on itself.")
+
+            # Check for duplicate dependencies
+            if len(dependencies) != len(set(dependencies)):
+                raise ValueError(
+                    f"Duplicate dependencies found for stage '{stage}'."
+                )
+
+        # Check for isolated nodes
+        # TODO: are there cases where isolated nodes would be valid for OneMod?
+        all_stages = set(self._stages.keys())
+        dependent_stages = set(
+            stage for deps in self._dependencies.values() for stage in deps
+        )
+        isolated_stages = (
+            all_stages - dependent_stages - set(self._dependencies.keys())
```

Comment:
I think every stage gets added to both self._stages and self._dependencies (possibly with an empty list) when added to the pipeline, so you might have to filter the last term by stages that actually have dependencies.

Reply:
Assuming this is all_stages - (stages that other stages depend on) - (stages that depend on other stages)